evm/vm/base.py

```diff
@@ -686,7 +673,52 @@ def validate_block(self, block):
                 )
             )
 
-    def validate_seal(self, header: BlockHeader) -> None:
+    @classmethod
+    def validate_header(cls, header: BlockHeader, parent_header: BlockHeader) -> None:
+        """
+        :raise evm.exceptions.ValidationError: if the header is not valid
+        """
+        validate_length_lte(header.extra_data, 32, title="BlockHeader.extra_data")
+
+        # Some validations are different for the genesis block
+        if header.is_genesis:
+            if parent_header is None:
+                # TODO validate a variety of things about the genesis block, like a matching:
+                # - mix_hash
+                # - difficulty
+                # - timestamp
+                # - coinbase
+                # - extra_data
+                # - gas_limit
+                pass
+            else:
+                raise ValidationError("Must not set parent header when validating genesis block")
+        else:
+            if parent_header is None:
+                raise ValidationError("Must set parent header if block is not genesis block")
+            else:
+                cls._validate_post_genesis_header(header, parent_header)
```

Comment:
Some thoughts on how to cleanup the readability here:

```python
if header.is_genesis and parent_header is not None:
    raise ...
elif parent_header is None and not header.is_genesis:
    raise ...
elif header.is_genesis:
    self._validate_genesis_header(...)
else:
    self._validate_post_genesis_header(...)
```

Reply:
#895 took this a fairly different direction and said `validate_header` is only to be used on non-genesis headers, which cleaned things up a lot.